File: src/main/java/com/glitch/knowledge/model/Station.java
package com.glitch.knowledge.model;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;

@Entity(name = "station")
public class Station {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(length = 128, nullable = false)
    private String name;

    @ManyToOne
    @JoinColumn(name = "tunnel_id", nullable = true)
    private TunnetNetwork tunnel;

    @ManyToMany()
    @JoinTable(name = "station_line", joinColumns = @JoinColumn(name = "station_id"), inverseJoinColumns = @JoinColumn(name = "line_id"))
    private Set<Line> connectingLines;

    @OneToMany(mappedBy = "station")
    private Set<StationToLine> lineConnectionInfo;

    /**
     * @return database ID
     */
    public int getId() {
        return this.id;
    }

    /**
     * @return Metro station name
     */
    public String getName() {
        return this.name;
    }

    /**
     * @return The {@link TunnelNetwork} if the station is part of one, otherwise
     *         null
     */
    public TunnetNetwork getTunnelNetwork() {
        return this.tunnel;
    }

    /**
     * @return {@link Line}s serviced by this station.
     */
    public Set<Line> getConnectingLines() {
        return this.connectingLines;
    }

    /**
     * Get the specific mapped data of the station as it relates to the given line.
     * 
     * @param line_id ID of {@link Line} serviced by station.
     * @return {@link StationToLine} object if line connects to station, otherwise
     *         null
     */
    public StationToLine getLineConnectionInfo(int line_id) {
        for (StationToLine connectionInfo : this.lineConnectionInfo) {
            if (connectionInfo.getLine().getId() == line_id)
                return connectionInfo;
        }
        return null;
    }

    /**
     * Get the other stations connected to the tunnel network that service the given
     * line.
     * 
     * @param line_id ID of {@link Line} to filter for.
     * @return List of valid {@link Station}s. Empty if there are no matches or if
     *         this station is not connected to a tunnel network.
     */
    public List<Station> getTunnelStationsConnectingToLine(int line_id) {
        List<Station> tunnelConnections = new ArrayList<Station>();

        if (this.tunnel == null)
            return tunnelConnections;

        for (Station tunnelStation : this.tunnel.getConnectingStations()) {
            if (tunnelStation.getId() == this.id)
                continue;

            for (Line line : tunnelStation.getConnectingLines()) {
                if (line.getId() == line_id) {
                    tunnelConnections.add(tunnelStation);
                    break;
                }
            }
        }

        return tunnelConnections;
    }
}

File: src/main/kotlin/com/glitch/knowledge/model/Station.kt
package com.glitch.knowledge.model

import org.jetbrains.exposed.v1.core.JoinType
import org.jetbrains.exposed.v1.core.Random
import org.jetbrains.exposed.v1.core.dao.id.EntityID
import org.jetbrains.exposed.v1.core.dao.id.IntIdTable
import org.jetbrains.exposed.v1.core.eq
import org.jetbrains.exposed.v1.core.neq
import org.jetbrains.exposed.v1.dao.ImmutableEntityClass
import org.jetbrains.exposed.v1.dao.IntEntity
import org.jetbrains.exposed.v1.dao.with
import org.jetbrains.exposed.v1.jdbc.andWhere
import org.jetbrains.exposed.v1.jdbc.select
import org.jetbrains.exposed.v1.jdbc.selectAll
import org.jetbrains.exposed.v1.jdbc.transactions.transaction


internal object StationTable : IntIdTable("station") {
    val name = varchar("name", 128)
    val tunnelId = optReference("tunnel_id", TunnelNetworkTable.id)
}


internal class StationDAO(id: EntityID<Int>) : IntEntity(id) {
    companion object : ImmutableEntityClass<Int, StationDAO>(StationTable)

    val name: String by StationTable.name
    val tunnelId by StationTable.tunnelId

    //val connectingLines by LineDAO.via(StationToLineTable.stationId, StationToLineTable.lineId)
    val lineConnectionInfo by StationToLineDAO.referrersOn(StationToLineTable.stationId)
}


/**
 * @property id Database ID.
 * @property name Name of the metro station.
 */
class Station(
    val id: Int,
    val name: String,
    private val tunnelId: EntityID<Int>?,
    private val connectingLinesInfo: List<StationToLine>
) {
    companion object {
        internal fun fromDAO(dao: StationDAO): Station {
            return Station(
                dao.id.value,
                dao.name,
                dao.tunnelId,
                transaction {
                    dao.lineConnectionInfo.with(StationDAO::lineConnectionInfo).map { StationToLine.fromDAO(it) }
                })
        }

        fun get(id: Int): Station? {
            return transaction {
                StationDAO.findById(id)?.let { fromDAO(it) }
            }
        }

        fun getTwoRandom(): List<Station> {
            return transaction {
                StationDAO.wrapRows(
                    StationTable.selectAll()
                        .orderBy(Random())
                        .limit(2)
                ).map { Station.fromDAO(it) }
            }

        }
    }

    fun isOnTunnel(): Boolean {
        return this.tunnelId != null
    }

    /**
     * Get the specific mapped data of the station as it relates to the given line.
     *
     * @param lineId ID of [Line] serviced by station.
     * @return [StationToLine] object if line connects to station, otherwise
     *         null
     */
    fun lineConnectionInfoFromLine(lineId: Int): StationToLine? {
        return this.connectingLinesInfo.find { it.lineId == lineId }
    }

    /**
     * Get the other stations connected to the tunnel network that service the given
     * line.
     *
     * @param lineId ID of [Line] to filter for.
     * @return List of valid [Station]. Empty if there are no matches or if
     *         this station is not connected to a tunnel network.
     */
    fun getTunnelStationsConnectingToLine(lineId: Int): List<Station> {
        val tunnelConnections: MutableList<Station> = mutableListOf()

        this.tunnelId ?: return tunnelConnections

        val stationId = this.id
        transaction {
            StationDAO.wrapRows(
                StationTable.join(
                    StationToLineTable,
                    JoinType.INNER,
                    onColumn = StationTable.id,
                    otherColumn = StationToLineTable.stationId
                ).select(StationTable.columns).andWhere {
                    StationTable.id neq stationId
                }.andWhere {
                    StationToLineTable.lineId eq lineId
                }.andWhere {
                    StationTable.tunnelId eq tunnelId
                }
            ).map { tunnelConnections.add(fromDAO(it)) }
        }

        return tunnelConnections
    }
}